{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "70b8d985",
   "metadata": {},
   "source": [
    "## We will use xena's API to retrieve our data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4be0704a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f1191bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import xenaPython as xena"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "57dfe6d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package xenaPython:\n",
      "\n",
      "NAME\n",
      "    xenaPython - Methods for querying data from UCSC Xena hubs\n",
      "\n",
      "DESCRIPTION\n",
      "    Data rows are associated with \"sample\" IDs.\n",
      "    Sample IDs are unique within a \"cohort\".\n",
      "    A \"dataset\" is a particular assay of a cohort, e.g. gene expression.\n",
      "    Datasets have associated metadata, specifying their data type and cohort.\n",
      "    \n",
      "    There are three primary data types: dense matrix (samples by probes),\n",
      "    sparse (sample, position, variant), and segmented (sample, position, value).\n",
      "    \n",
      "    \n",
      "    Dense matrices can be genotypic or phenotypic. Phenotypic matrices have\n",
      "    associated field metadata (descriptive names, codes, etc.).\n",
      "    \n",
      "    Genotypic matricies may have an associated probeMap, which maps probes to\n",
      "    genomic locations. If a matrix has hugo probeMap, the probes themselves\n",
      "    are gene names. Otherwise, a probeMap is used to map a gene location to a\n",
      "    set of probes.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    convert\n",
      "    example\n",
      "    jupyter\n",
      "    xenaAPI\n",
      "    xenaQuery\n",
      "\n",
      "FUNCTIONS\n",
      "    all_cohorts(host, exclude)\n",
      "    \n",
      "    all_datasets(host)\n",
      "    \n",
      "    all_datasets_n(host)\n",
      "        Count the number datasets with non-null cohort\n",
      "    \n",
      "    all_field_metadata(host, dataset)\n",
      "        Metadata for all dataset fields (phenotypic datasets)\n",
      "    \n",
      "    cohort_samples(host, cohort, limit)\n",
      "        All samples in cohort\n",
      "    \n",
      "    cohort_summary(host, exclude)\n",
      "        Count datasets per-cohort, excluding the given dataset types\n",
      "        \n",
      "        xena.cohort_summary(xena.PUBLIC_HUBS[\"pancanAtlasHub\"], [\"probeMap\"])\n",
      "    \n",
      "    dataset_fetch(host, dataset, samples, probes)\n",
      "        Probe values for give samples\n",
      "    \n",
      "    dataset_field(host, dataset)\n",
      "        All field (probe) names in dataset\n",
      "    \n",
      "    dataset_field_examples(host, dataset, count)\n",
      "        Field names in dataset, up to <count>\n",
      "    \n",
      "    dataset_field_n(host, dataset)\n",
      "        Number of fields in dataset\n",
      "    \n",
      "    dataset_gene_probe_avg(host, dataset, samples, genes)\n",
      "        Probe average, per-gene, for given samples\n",
      "    \n",
      "    dataset_gene_probes_values(host, dataset, samples, genes)\n",
      "        Probe values in gene, and probe genomic positions, for given samples\n",
      "    \n",
      "    dataset_list(host, cohorts)\n",
      "        Dataset metadata for datasets in the given cohorts\n",
      "    \n",
      "    dataset_metadata(host, dataset)\n",
      "        Dataset metadata\n",
      "    \n",
      "    dataset_probe_signature(host, dataset, samples, probes, weights)\n",
      "        Computed probe signature for given samples and weight array\n",
      "    \n",
      "    dataset_probe_values(host, dataset, samples, probes)\n",
      "        Probe values for given samples, and probe genomic positions\n",
      "        \n",
      "        host = xena.PUBLIC_HUBS[\"pancanAtlasHub\"]\n",
      "        dataset = \"EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena\"\n",
      "        samples = xena.dataset_samples(host, dataset, None)\n",
      "        [position, [foxm1, tp53]] = xena.dataset_probe_values(host, dataset, samples, [\"FOXM1\", \"TP53\"])\n",
      "    \n",
      "    dataset_samples(host, dataset, limit)\n",
      "        All samples in dataset (optional limit)\n",
      "        \n",
      "        samples = xena.dataset_samples(xena.PUBLIC_HUBS[\"pancanAtlasHub\"], \"EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena\", None)\n",
      "    \n",
      "    dataset_samples_n_dense_matrix(host, dataset)\n",
      "        All samples in dataset (faster, for dense matrix dataset only)\n",
      "    \n",
      "    dataset_sources(host, dataset)\n",
      "    \n",
      "    dataset_status(host, dataset)\n",
      "    \n",
      "    datasets_null_rows(host)\n",
      "    \n",
      "    feature_list(host, dataset)\n",
      "        Dataset field names and long titles (phenotypic datasets)\n",
      "    \n",
      "    field_codes(host, dataset, fields)\n",
      "        Codes for categorical fields\n",
      "    \n",
      "    field_metadata(host, dataset, fields)\n",
      "        Metadata for given fields (phenotypic datasets)\n",
      "    \n",
      "    gene_transcripts(host, dataset, gene)\n",
      "        Gene transcripts\n",
      "    \n",
      "    load_file(name, host='http://127.0.0.1:7222')\n",
      "        # XXX need timeout and error handling\n",
      "        # should this do the file copy, too?\n",
      "    \n",
      "    load_ipython_extension(ipython)\n",
      "        jupyter support method\n",
      "    \n",
      "    match_fields(host, dataset, names)\n",
      "        Find fields matching names (must be lower-case)\n",
      "    \n",
      "    open_browser(url='heatmap/', columns=[])\n",
      "    \n",
      "    probe_count(host, dataset)\n",
      "    \n",
      "    probemap_list(host)\n",
      "        Find probemaps\n",
      "    \n",
      "    ref_gene_exons(host, dataset, genes)\n",
      "        Gene model\n",
      "    \n",
      "    ref_gene_position(host, dataset, gene)\n",
      "        Gene position from gene model\n",
      "    \n",
      "    ref_gene_range(host, dataset, chr, start, end)\n",
      "        Gene models overlapping range\n",
      "    \n",
      "    segment_data_examples(host, dataset, count)\n",
      "        Initial segmented data rows, with limit\n",
      "    \n",
      "    segmented_data_range(host, dataset, samples, chr, start, end)\n",
      "        Segmented (copy number) data overlapping range\n",
      "    \n",
      "    sparse_data(host, dataset, samples, genes)\n",
      "        Sparse (mutation) data rows for genes\n",
      "    \n",
      "    sparse_data_examples(host, dataset, count)\n",
      "        Initial sparse data rows, with limit\n",
      "    \n",
      "    sparse_data_match_field(host, field, dataset, names)\n",
      "        Genes in sparse (mutation) dataset matching given names\n",
      "    \n",
      "    sparse_data_match_field_slow(host, field, dataset, names)\n",
      "        Genes in sparse (mutation) dataset matching given names, case-insensitive (names must be lower-case)\n",
      "    \n",
      "    sparse_data_match_partial_field(host, field, dataset, names, limit)\n",
      "        Partial match genes in sparse (mutation) dataset\n",
      "    \n",
      "    sparse_data_range(host, dataset, samples, chr, start, end)\n",
      "        Sparse (mutation) data rows overlapping the given range, for the given samples\n",
      "    \n",
      "    transcript_expression(host, transcripts, studyA, subtypeA, studyB, subtypeB, dataset)\n",
      "    \n",
      "    unload_ipython_extension(ipython)\n",
      "        jupyter support method\n",
      "\n",
      "DATA\n",
      "    LOCAL_HUB = 'https://local.xena.ucsc.edu:7223'\n",
      "    PUBLIC_HUBS = {'gdcHub': 'https://gdc.xenahubs.net', 'icgcHub': 'https...\n",
      "    QUERIES = {'all_cohorts': ';allCohorts\\n(fn [exclude]\\n\\t(map :cohort\\...\n",
      "    __warningregistry__ = {'version': 17, ('Possible nested set at positio...\n",
      "    excludeType = ['probeMap', 'probemap', 'genePredExt']\n",
      "    ipython_instance = None\n",
      "\n",
      "FILE\n",
      "    /home/ubuntu/anaconda3/lib/python3.7/site-packages/xenaPython/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(xena)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d62587d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "host = 'https://toil.xenahubs.net'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa14a59b",
   "metadata": {},
   "source": [
    "### Our dataset is the RSEM expected_count (DESeq2 standardized) \n",
    "### i.e. 'TCGA-GTEx-TARGET-gene-exp-counts.deseq2-normalized.log2' from the above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8ec85fbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = 'TCGA-GTEx-TARGET-gene-exp-counts.deseq2-normalized.log2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "78efcf23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[19039]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xena.dataset_samples_n_dense_matrix(host, dataset) #That is the full number of samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b48964ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = xena.dataset_samples(host, dataset,19039) # All of the samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9769fff1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19039"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7b87c2f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TCGA-DX-A7EO-01'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3220fd9a",
   "metadata": {},
   "source": [
    "### We will split our data into TCGA , TARGET and GTEX samples. This will be done in order for us to be able to load datasets as well as make the analysis better"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "15f2294a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tcga = []\n",
    "target = []\n",
    "gtex = []\n",
    "\n",
    "for i in range(len(samples)):\n",
    "    if samples[i].startswith('TCGA'):\n",
    "        tcga.append(samples[i])\n",
    "    elif samples[i].startswith('TARGET'):\n",
    "        target.append(samples[i])\n",
    "    else:\n",
    "        gtex.append(samples[i])\n",
    "            \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f7fe3200",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10530"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tcga)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "57271d31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "734"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "03faf12e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7775"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gtex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8183169a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = xena.dataset_field(host, dataset) #all of the genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e1695d54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60499"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f1d05656",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Let's create a function that we pass batches of genes and samples and creates a dataframe \n",
    "\n",
    "def pos(samples, genes):\n",
    "    cohort = 'TCGA TARGET GTEx'\n",
    "    host = 'https://toil.xenahubs.net'\n",
    "    dataset = 'TCGA-GTEx-TARGET-gene-exp-counts.deseq2-normalized.log2'\n",
    "    \n",
    "    positions = xena.dataset_probe_values(host, dataset, samples, genes) #this returns the normalized position of the genes\n",
    "    gene_pos = dict(zip(genes, positions[1]))\n",
    "    df = pd.DataFrame.from_dict(gene_pos)\n",
    "    df.index = samples\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "af846a7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "chunks = [genes[x:x+1000] for x in range(0, len(genes), 1000)] #let's take chuncks of 1000 genes at a time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f18c3cd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "61"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(chunks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "284c25d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pos(samples, genes[9800:9900])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d49a51bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000154719.13</th>\n",
       "      <th>ENSG00000154721.14</th>\n",
       "      <th>ENSG00000154723.12</th>\n",
       "      <th>ENSG00000154727.10</th>\n",
       "      <th>ENSG00000154734.14</th>\n",
       "      <th>ENSG00000154736.5</th>\n",
       "      <th>ENSG00000154743.17</th>\n",
       "      <th>ENSG00000154760.13</th>\n",
       "      <th>ENSG00000154764.5</th>\n",
       "      <th>ENSG00000154767.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000155629.14</th>\n",
       "      <th>ENSG00000155636.14</th>\n",
       "      <th>ENSG00000155640.6</th>\n",
       "      <th>ENSG00000155657.23</th>\n",
       "      <th>ENSG00000155659.14</th>\n",
       "      <th>ENSG00000155660.10</th>\n",
       "      <th>ENSG00000155666.11</th>\n",
       "      <th>ENSG00000155714.13</th>\n",
       "      <th>ENSG00000155719.16</th>\n",
       "      <th>ENSG00000155729.12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-AD-5900-01</th>\n",
       "      <td>10.100</td>\n",
       "      <td>8.004</td>\n",
       "      <td>11.78</td>\n",
       "      <td>10.34</td>\n",
       "      <td>9.243</td>\n",
       "      <td>6.769</td>\n",
       "      <td>8.903</td>\n",
       "      <td>9.766</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>10.950</td>\n",
       "      <td>...</td>\n",
       "      <td>11.080</td>\n",
       "      <td>8.503</td>\n",
       "      <td>8.026</td>\n",
       "      <td>6.058</td>\n",
       "      <td>9.794</td>\n",
       "      <td>13.72</td>\n",
       "      <td>7.234</td>\n",
       "      <td>1.551</td>\n",
       "      <td>3.413</td>\n",
       "      <td>9.271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BP-4968-01</th>\n",
       "      <td>10.600</td>\n",
       "      <td>9.964</td>\n",
       "      <td>12.56</td>\n",
       "      <td>11.11</td>\n",
       "      <td>12.880</td>\n",
       "      <td>9.522</td>\n",
       "      <td>8.345</td>\n",
       "      <td>11.270</td>\n",
       "      <td>2.3690</td>\n",
       "      <td>11.260</td>\n",
       "      <td>...</td>\n",
       "      <td>11.440</td>\n",
       "      <td>9.026</td>\n",
       "      <td>6.116</td>\n",
       "      <td>8.028</td>\n",
       "      <td>9.687</td>\n",
       "      <td>13.67</td>\n",
       "      <td>8.549</td>\n",
       "      <td>2.116</td>\n",
       "      <td>6.310</td>\n",
       "      <td>10.040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-NG-A4VU-01</th>\n",
       "      <td>10.210</td>\n",
       "      <td>10.730</td>\n",
       "      <td>12.82</td>\n",
       "      <td>11.04</td>\n",
       "      <td>11.350</td>\n",
       "      <td>10.420</td>\n",
       "      <td>8.149</td>\n",
       "      <td>7.096</td>\n",
       "      <td>5.5670</td>\n",
       "      <td>9.727</td>\n",
       "      <td>...</td>\n",
       "      <td>9.689</td>\n",
       "      <td>9.037</td>\n",
       "      <td>7.748</td>\n",
       "      <td>16.150</td>\n",
       "      <td>8.179</td>\n",
       "      <td>12.84</td>\n",
       "      <td>7.564</td>\n",
       "      <td>2.150</td>\n",
       "      <td>4.545</td>\n",
       "      <td>8.947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CG-4305-01</th>\n",
       "      <td>10.200</td>\n",
       "      <td>8.962</td>\n",
       "      <td>11.79</td>\n",
       "      <td>11.11</td>\n",
       "      <td>10.100</td>\n",
       "      <td>8.454</td>\n",
       "      <td>10.140</td>\n",
       "      <td>10.170</td>\n",
       "      <td>1.4770</td>\n",
       "      <td>10.730</td>\n",
       "      <td>...</td>\n",
       "      <td>11.320</td>\n",
       "      <td>7.934</td>\n",
       "      <td>9.116</td>\n",
       "      <td>8.587</td>\n",
       "      <td>11.260</td>\n",
       "      <td>13.80</td>\n",
       "      <td>7.140</td>\n",
       "      <td>1.878</td>\n",
       "      <td>4.302</td>\n",
       "      <td>9.577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AO-A03M-01</th>\n",
       "      <td>9.494</td>\n",
       "      <td>9.442</td>\n",
       "      <td>11.82</td>\n",
       "      <td>10.82</td>\n",
       "      <td>11.680</td>\n",
       "      <td>8.287</td>\n",
       "      <td>7.884</td>\n",
       "      <td>8.230</td>\n",
       "      <td>0.6768</td>\n",
       "      <td>10.990</td>\n",
       "      <td>...</td>\n",
       "      <td>10.400</td>\n",
       "      <td>8.733</td>\n",
       "      <td>9.143</td>\n",
       "      <td>7.361</td>\n",
       "      <td>9.391</td>\n",
       "      <td>14.79</td>\n",
       "      <td>7.638</td>\n",
       "      <td>0.000</td>\n",
       "      <td>4.533</td>\n",
       "      <td>9.623</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ENSG00000154719.13  ENSG00000154721.14  ENSG00000154723.12  \\\n",
       "TCGA-AD-5900-01              10.100               8.004               11.78   \n",
       "TCGA-BP-4968-01              10.600               9.964               12.56   \n",
       "TCGA-NG-A4VU-01              10.210              10.730               12.82   \n",
       "TCGA-CG-4305-01              10.200               8.962               11.79   \n",
       "TCGA-AO-A03M-01               9.494               9.442               11.82   \n",
       "\n",
       "                 ENSG00000154727.10  ENSG00000154734.14  ENSG00000154736.5  \\\n",
       "TCGA-AD-5900-01               10.34               9.243              6.769   \n",
       "TCGA-BP-4968-01               11.11              12.880              9.522   \n",
       "TCGA-NG-A4VU-01               11.04              11.350             10.420   \n",
       "TCGA-CG-4305-01               11.11              10.100              8.454   \n",
       "TCGA-AO-A03M-01               10.82              11.680              8.287   \n",
       "\n",
       "                 ENSG00000154743.17  ENSG00000154760.13  ENSG00000154764.5  \\\n",
       "TCGA-AD-5900-01               8.903               9.766             0.0000   \n",
       "TCGA-BP-4968-01               8.345              11.270             2.3690   \n",
       "TCGA-NG-A4VU-01               8.149               7.096             5.5670   \n",
       "TCGA-CG-4305-01              10.140              10.170             1.4770   \n",
       "TCGA-AO-A03M-01               7.884               8.230             0.6768   \n",
       "\n",
       "                 ENSG00000154767.14  ...  ENSG00000155629.14  \\\n",
       "TCGA-AD-5900-01              10.950  ...              11.080   \n",
       "TCGA-BP-4968-01              11.260  ...              11.440   \n",
       "TCGA-NG-A4VU-01               9.727  ...               9.689   \n",
       "TCGA-CG-4305-01              10.730  ...              11.320   \n",
       "TCGA-AO-A03M-01              10.990  ...              10.400   \n",
       "\n",
       "                 ENSG00000155636.14  ENSG00000155640.6  ENSG00000155657.23  \\\n",
       "TCGA-AD-5900-01               8.503              8.026               6.058   \n",
       "TCGA-BP-4968-01               9.026              6.116               8.028   \n",
       "TCGA-NG-A4VU-01               9.037              7.748              16.150   \n",
       "TCGA-CG-4305-01               7.934              9.116               8.587   \n",
       "TCGA-AO-A03M-01               8.733              9.143               7.361   \n",
       "\n",
       "                 ENSG00000155659.14  ENSG00000155660.10  ENSG00000155666.11  \\\n",
       "TCGA-AD-5900-01               9.794               13.72               7.234   \n",
       "TCGA-BP-4968-01               9.687               13.67               8.549   \n",
       "TCGA-NG-A4VU-01               8.179               12.84               7.564   \n",
       "TCGA-CG-4305-01              11.260               13.80               7.140   \n",
       "TCGA-AO-A03M-01               9.391               14.79               7.638   \n",
       "\n",
       "                 ENSG00000155714.13  ENSG00000155719.16  ENSG00000155729.12  \n",
       "TCGA-AD-5900-01               1.551               3.413               9.271  \n",
       "TCGA-BP-4968-01               2.116               6.310              10.040  \n",
       "TCGA-NG-A4VU-01               2.150               4.545               8.947  \n",
       "TCGA-CG-4305-01               1.878               4.302               9.577  \n",
       "TCGA-AO-A03M-01               0.000               4.533               9.623  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4cbf2298",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000154719.13</th>\n",
       "      <th>ENSG00000154721.14</th>\n",
       "      <th>ENSG00000154723.12</th>\n",
       "      <th>ENSG00000154727.10</th>\n",
       "      <th>ENSG00000154734.14</th>\n",
       "      <th>ENSG00000154736.5</th>\n",
       "      <th>ENSG00000154743.17</th>\n",
       "      <th>ENSG00000154760.13</th>\n",
       "      <th>ENSG00000154764.5</th>\n",
       "      <th>ENSG00000154767.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000155629.14</th>\n",
       "      <th>ENSG00000155636.14</th>\n",
       "      <th>ENSG00000155640.6</th>\n",
       "      <th>ENSG00000155657.23</th>\n",
       "      <th>ENSG00000155659.14</th>\n",
       "      <th>ENSG00000155660.10</th>\n",
       "      <th>ENSG00000155666.11</th>\n",
       "      <th>ENSG00000155714.13</th>\n",
       "      <th>ENSG00000155719.16</th>\n",
       "      <th>ENSG00000155729.12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "      <td>19039.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>9.216853</td>\n",
       "      <td>8.893895</td>\n",
       "      <td>11.423807</td>\n",
       "      <td>10.287863</td>\n",
       "      <td>11.049464</td>\n",
       "      <td>7.910874</td>\n",
       "      <td>8.506836</td>\n",
       "      <td>8.472479</td>\n",
       "      <td>3.440166</td>\n",
       "      <td>10.856846</td>\n",
       "      <td>...</td>\n",
       "      <td>8.838714</td>\n",
       "      <td>8.388410</td>\n",
       "      <td>7.110810</td>\n",
       "      <td>8.789456</td>\n",
       "      <td>8.731416</td>\n",
       "      <td>12.326441</td>\n",
       "      <td>7.433425</td>\n",
       "      <td>1.930716</td>\n",
       "      <td>3.616432</td>\n",
       "      <td>9.241333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.964574</td>\n",
       "      <td>1.957889</td>\n",
       "      <td>1.098617</td>\n",
       "      <td>0.927541</td>\n",
       "      <td>2.353984</td>\n",
       "      <td>2.382528</td>\n",
       "      <td>0.823556</td>\n",
       "      <td>2.166364</td>\n",
       "      <td>3.372494</td>\n",
       "      <td>0.843610</td>\n",
       "      <td>...</td>\n",
       "      <td>2.491209</td>\n",
       "      <td>0.737831</td>\n",
       "      <td>1.353833</td>\n",
       "      <td>2.617947</td>\n",
       "      <td>2.492504</td>\n",
       "      <td>1.839742</td>\n",
       "      <td>1.043842</td>\n",
       "      <td>1.407380</td>\n",
       "      <td>1.957262</td>\n",
       "      <td>0.865747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>8.505000</td>\n",
       "      <td>7.790000</td>\n",
       "      <td>10.720000</td>\n",
       "      <td>9.663500</td>\n",
       "      <td>9.874500</td>\n",
       "      <td>6.663000</td>\n",
       "      <td>8.085000</td>\n",
       "      <td>6.862500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.390000</td>\n",
       "      <td>...</td>\n",
       "      <td>7.031000</td>\n",
       "      <td>7.954000</td>\n",
       "      <td>6.107000</td>\n",
       "      <td>7.493000</td>\n",
       "      <td>7.307500</td>\n",
       "      <td>10.830000</td>\n",
       "      <td>6.731000</td>\n",
       "      <td>1.007500</td>\n",
       "      <td>2.302000</td>\n",
       "      <td>8.673000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>9.286000</td>\n",
       "      <td>9.085000</td>\n",
       "      <td>11.480000</td>\n",
       "      <td>10.220000</td>\n",
       "      <td>11.270000</td>\n",
       "      <td>8.174000</td>\n",
       "      <td>8.472000</td>\n",
       "      <td>8.499000</td>\n",
       "      <td>2.330000</td>\n",
       "      <td>10.850000</td>\n",
       "      <td>...</td>\n",
       "      <td>8.911000</td>\n",
       "      <td>8.374000</td>\n",
       "      <td>6.956000</td>\n",
       "      <td>8.524000</td>\n",
       "      <td>8.887000</td>\n",
       "      <td>12.510000</td>\n",
       "      <td>7.401000</td>\n",
       "      <td>1.824000</td>\n",
       "      <td>3.660000</td>\n",
       "      <td>9.203000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>9.890000</td>\n",
       "      <td>10.210000</td>\n",
       "      <td>12.120000</td>\n",
       "      <td>10.930000</td>\n",
       "      <td>12.605000</td>\n",
       "      <td>9.446000</td>\n",
       "      <td>8.954000</td>\n",
       "      <td>10.170000</td>\n",
       "      <td>6.031000</td>\n",
       "      <td>11.320000</td>\n",
       "      <td>...</td>\n",
       "      <td>10.600000</td>\n",
       "      <td>8.873000</td>\n",
       "      <td>8.010000</td>\n",
       "      <td>9.362000</td>\n",
       "      <td>10.410000</td>\n",
       "      <td>13.820000</td>\n",
       "      <td>8.047000</td>\n",
       "      <td>2.681000</td>\n",
       "      <td>4.909000</td>\n",
       "      <td>9.866000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>13.100000</td>\n",
       "      <td>15.510000</td>\n",
       "      <td>16.740000</td>\n",
       "      <td>13.990000</td>\n",
       "      <td>17.040000</td>\n",
       "      <td>16.480000</td>\n",
       "      <td>12.850000</td>\n",
       "      <td>14.840000</td>\n",
       "      <td>15.420000</td>\n",
       "      <td>14.350000</td>\n",
       "      <td>...</td>\n",
       "      <td>16.520000</td>\n",
       "      <td>10.800000</td>\n",
       "      <td>12.450000</td>\n",
       "      <td>24.160000</td>\n",
       "      <td>16.810000</td>\n",
       "      <td>17.530000</td>\n",
       "      <td>13.910000</td>\n",
       "      <td>9.934000</td>\n",
       "      <td>12.350000</td>\n",
       "      <td>12.490000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ENSG00000154719.13  ENSG00000154721.14  ENSG00000154723.12  \\\n",
       "count        19039.000000        19039.000000        19039.000000   \n",
       "mean             9.216853            8.893895           11.423807   \n",
       "std              0.964574            1.957889            1.098617   \n",
       "min              0.000000            0.000000            0.000000   \n",
       "25%              8.505000            7.790000           10.720000   \n",
       "50%              9.286000            9.085000           11.480000   \n",
       "75%              9.890000           10.210000           12.120000   \n",
       "max             13.100000           15.510000           16.740000   \n",
       "\n",
       "       ENSG00000154727.10  ENSG00000154734.14  ENSG00000154736.5  \\\n",
       "count        19039.000000        19039.000000       19039.000000   \n",
       "mean            10.287863           11.049464           7.910874   \n",
       "std              0.927541            2.353984           2.382528   \n",
       "min              0.000000            0.000000           0.000000   \n",
       "25%              9.663500            9.874500           6.663000   \n",
       "50%             10.220000           11.270000           8.174000   \n",
       "75%             10.930000           12.605000           9.446000   \n",
       "max             13.990000           17.040000          16.480000   \n",
       "\n",
       "       ENSG00000154743.17  ENSG00000154760.13  ENSG00000154764.5  \\\n",
       "count        19039.000000        19039.000000       19039.000000   \n",
       "mean             8.506836            8.472479           3.440166   \n",
       "std              0.823556            2.166364           3.372494   \n",
       "min              0.000000            0.000000           0.000000   \n",
       "25%              8.085000            6.862500           0.000000   \n",
       "50%              8.472000            8.499000           2.330000   \n",
       "75%              8.954000           10.170000           6.031000   \n",
       "max             12.850000           14.840000          15.420000   \n",
       "\n",
       "       ENSG00000154767.14  ...  ENSG00000155629.14  ENSG00000155636.14  \\\n",
       "count        19039.000000  ...        19039.000000        19039.000000   \n",
       "mean            10.856846  ...            8.838714            8.388410   \n",
       "std              0.843610  ...            2.491209            0.737831   \n",
       "min              0.000000  ...            0.000000            0.000000   \n",
       "25%             10.390000  ...            7.031000            7.954000   \n",
       "50%             10.850000  ...            8.911000            8.374000   \n",
       "75%             11.320000  ...           10.600000            8.873000   \n",
       "max             14.350000  ...           16.520000           10.800000   \n",
       "\n",
       "       ENSG00000155640.6  ENSG00000155657.23  ENSG00000155659.14  \\\n",
       "count       19039.000000        19039.000000        19039.000000   \n",
       "mean            7.110810            8.789456            8.731416   \n",
       "std             1.353833            2.617947            2.492504   \n",
       "min             0.000000            0.000000            0.000000   \n",
       "25%             6.107000            7.493000            7.307500   \n",
       "50%             6.956000            8.524000            8.887000   \n",
       "75%             8.010000            9.362000           10.410000   \n",
       "max            12.450000           24.160000           16.810000   \n",
       "\n",
       "       ENSG00000155660.10  ENSG00000155666.11  ENSG00000155714.13  \\\n",
       "count        19039.000000        19039.000000        19039.000000   \n",
       "mean            12.326441            7.433425            1.930716   \n",
       "std              1.839742            1.043842            1.407380   \n",
       "min              0.000000            0.000000            0.000000   \n",
       "25%             10.830000            6.731000            1.007500   \n",
       "50%             12.510000            7.401000            1.824000   \n",
       "75%             13.820000            8.047000            2.681000   \n",
       "max             17.530000           13.910000            9.934000   \n",
       "\n",
       "       ENSG00000155719.16  ENSG00000155729.12  \n",
       "count        19039.000000        19039.000000  \n",
       "mean             3.616432            9.241333  \n",
       "std              1.957262            0.865747  \n",
       "min              0.000000            0.000000  \n",
       "25%              2.302000            8.673000  \n",
       "50%              3.660000            9.203000  \n",
       "75%              4.909000            9.866000  \n",
       "max             12.350000           12.490000  \n",
       "\n",
       "[8 rows x 100 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08297a8c",
   "metadata": {},
   "source": [
    "### We will work with batches of 1000 features. According to:\n",
    "### Jaskowiak, P. A., Costa, I. G., & Campello, R. J. G. B. (2018). Clustering of RNA-Seq samples: Comparison study on cancer data. Methods, 132, 42–49. https://doi.org/10.1016/j.ymeth.2017.07.023 ,\n",
    "### using samples with 1000 features at a time, can lead to superior results in clustering. We will begin our analysis with datasets of 1000 features at a time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "608cecfc",
   "metadata": {},
   "source": [
    "### We will focus our analysis in tcga samples for the time being\n",
    "\n",
    "### According to Vidman, L., Källberg, D., &#38; Rydén, P. (2019). Cluster analysis on high dimensional RNA-seq data with applications to cancer research - An evaluation study. PLoS ONE, 1412). https://doi.org/10.1371/journal.pone.0219102 genes  that either showed\n",
    "### 1)  small variation between samples (i.e. standard deviation less than  0.5) \n",
    "### 2) or were 0 everywhere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "80581940",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function for use if we decide to get all of the data\n",
    "nums = list(range(0,61,1))\n",
    "\n",
    "names = ['df'+str(nums[x]) for x in range(len(nums))]\n",
    "\n",
    "\n",
    "for i in range(len(names)):\n",
    "    names[i] = pos(tcga, chunks[i])\n",
    "    names[i] = names[i].loc[:, (names[i]!= 0).any(axis=0)]\n",
    "    names[i].drop(names[i].std()[names[i].std() < 0.5].index.values, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "dee177d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = []\n",
    "for i in range(len(names)):\n",
    "    dataframes.append(names[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b9a067f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.concat([df.set_index(df.index) for df in dataframes], ignore_index=False, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "22596c39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 10530 entries, TCGA-AD-5900-01 to TCGA-FU-A3HZ-01\n",
      "Columns: 57060 entries, ENSG00000000003.14 to sampleID\n",
      "dtypes: float64(57059), int64(1)\n",
      "memory usage: 4.5+ GB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a6eefc65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <th>ENSG00000000938.12</th>\n",
       "      <th>ENSG00000000971.15</th>\n",
       "      <th>ENSG00000001036.13</th>\n",
       "      <th>ENSG00000001084.10</th>\n",
       "      <th>ENSG00000001167.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000282785.1</th>\n",
       "      <th>ENSG00000282787.1</th>\n",
       "      <th>ENSG00000282793.1</th>\n",
       "      <th>ENSG00000282795.1</th>\n",
       "      <th>ENSG00000282798.1</th>\n",
       "      <th>ENSG00000282804.1</th>\n",
       "      <th>ENSG00000282807.1</th>\n",
       "      <th>ENSG00000282815.1</th>\n",
       "      <th>ENSG00000282816.1</th>\n",
       "      <th>sampleID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-AD-5900-01</th>\n",
       "      <td>10.670</td>\n",
       "      <td>0.000</td>\n",
       "      <td>10.38</td>\n",
       "      <td>9.036</td>\n",
       "      <td>8.496</td>\n",
       "      <td>8.563</td>\n",
       "      <td>10.050</td>\n",
       "      <td>11.44</td>\n",
       "      <td>10.860</td>\n",
       "      <td>9.896</td>\n",
       "      <td>...</td>\n",
       "      <td>1.551</td>\n",
       "      <td>1.962</td>\n",
       "      <td>0.9745</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3.538</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BP-4968-01</th>\n",
       "      <td>11.580</td>\n",
       "      <td>5.298</td>\n",
       "      <td>10.51</td>\n",
       "      <td>9.421</td>\n",
       "      <td>7.842</td>\n",
       "      <td>10.280</td>\n",
       "      <td>12.410</td>\n",
       "      <td>12.69</td>\n",
       "      <td>10.940</td>\n",
       "      <td>10.290</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.116</td>\n",
       "      <td>0.8746</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.773</td>\n",
       "      <td>0.8746</td>\n",
       "      <td>0.8746</td>\n",
       "      <td>0.8746</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-NG-A4VU-01</th>\n",
       "      <td>9.693</td>\n",
       "      <td>3.262</td>\n",
       "      <td>11.00</td>\n",
       "      <td>9.155</td>\n",
       "      <td>9.683</td>\n",
       "      <td>7.383</td>\n",
       "      <td>8.848</td>\n",
       "      <td>11.19</td>\n",
       "      <td>9.806</td>\n",
       "      <td>11.180</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.977</td>\n",
       "      <td>3.2620</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.977</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CG-4305-01</th>\n",
       "      <td>10.260</td>\n",
       "      <td>0.000</td>\n",
       "      <td>11.10</td>\n",
       "      <td>9.465</td>\n",
       "      <td>9.078</td>\n",
       "      <td>9.173</td>\n",
       "      <td>12.320</td>\n",
       "      <td>12.09</td>\n",
       "      <td>12.220</td>\n",
       "      <td>11.040</td>\n",
       "      <td>...</td>\n",
       "      <td>6.791</td>\n",
       "      <td>1.878</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>6.066</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AO-A03M-01</th>\n",
       "      <td>10.100</td>\n",
       "      <td>1.997</td>\n",
       "      <td>11.74</td>\n",
       "      <td>10.380</td>\n",
       "      <td>9.157</td>\n",
       "      <td>8.509</td>\n",
       "      <td>11.630</td>\n",
       "      <td>11.80</td>\n",
       "      <td>10.580</td>\n",
       "      <td>11.680</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3.230</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.136</td>\n",
       "      <td>4.050</td>\n",
       "      <td>0.6768</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3.8250</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 57060 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ENSG00000000003.14  ENSG00000000005.5  ENSG00000000419.12  \\\n",
       "TCGA-AD-5900-01              10.670              0.000               10.38   \n",
       "TCGA-BP-4968-01              11.580              5.298               10.51   \n",
       "TCGA-NG-A4VU-01               9.693              3.262               11.00   \n",
       "TCGA-CG-4305-01              10.260              0.000               11.10   \n",
       "TCGA-AO-A03M-01              10.100              1.997               11.74   \n",
       "\n",
       "                 ENSG00000000457.13  ENSG00000000460.16  ENSG00000000938.12  \\\n",
       "TCGA-AD-5900-01               9.036               8.496               8.563   \n",
       "TCGA-BP-4968-01               9.421               7.842              10.280   \n",
       "TCGA-NG-A4VU-01               9.155               9.683               7.383   \n",
       "TCGA-CG-4305-01               9.465               9.078               9.173   \n",
       "TCGA-AO-A03M-01              10.380               9.157               8.509   \n",
       "\n",
       "                 ENSG00000000971.15  ENSG00000001036.13  ENSG00000001084.10  \\\n",
       "TCGA-AD-5900-01              10.050               11.44              10.860   \n",
       "TCGA-BP-4968-01              12.410               12.69              10.940   \n",
       "TCGA-NG-A4VU-01               8.848               11.19               9.806   \n",
       "TCGA-CG-4305-01              12.320               12.09              12.220   \n",
       "TCGA-AO-A03M-01              11.630               11.80              10.580   \n",
       "\n",
       "                 ENSG00000001167.14  ...  ENSG00000282785.1  \\\n",
       "TCGA-AD-5900-01               9.896  ...              1.551   \n",
       "TCGA-BP-4968-01              10.290  ...              0.000   \n",
       "TCGA-NG-A4VU-01              11.180  ...              0.000   \n",
       "TCGA-CG-4305-01              11.040  ...              6.791   \n",
       "TCGA-AO-A03M-01              11.680  ...              0.000   \n",
       "\n",
       "                 ENSG00000282787.1  ENSG00000282793.1  ENSG00000282795.1  \\\n",
       "TCGA-AD-5900-01              1.962             0.9745              0.000   \n",
       "TCGA-BP-4968-01              2.116             0.8746              0.000   \n",
       "TCGA-NG-A4VU-01              2.977             3.2620              0.000   \n",
       "TCGA-CG-4305-01              1.878             0.0000              0.000   \n",
       "TCGA-AO-A03M-01              3.230             0.0000              1.136   \n",
       "\n",
       "                 ENSG00000282798.1  ENSG00000282804.1  ENSG00000282807.1  \\\n",
       "TCGA-AD-5900-01              3.538             0.0000             0.0000   \n",
       "TCGA-BP-4968-01              2.773             0.8746             0.8746   \n",
       "TCGA-NG-A4VU-01              2.977             0.0000             0.0000   \n",
       "TCGA-CG-4305-01              6.066             0.0000             0.0000   \n",
       "TCGA-AO-A03M-01              4.050             0.6768             0.0000   \n",
       "\n",
       "                 ENSG00000282815.1  ENSG00000282816.1  sampleID  \n",
       "TCGA-AD-5900-01             0.0000                0.0         0  \n",
       "TCGA-BP-4968-01             0.8746                0.0         1  \n",
       "TCGA-NG-A4VU-01             0.0000                0.0         2  \n",
       "TCGA-CG-4305-01             0.0000                0.0         3  \n",
       "TCGA-AO-A03M-01             3.8250                0.0         4  \n",
       "\n",
       "[5 rows x 57060 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "dccccc8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10530"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3ae43038",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 10530 entries, TCGA-AD-5900-01 to TCGA-FU-A3HZ-01\n",
      "Columns: 57060 entries, ENSG00000000003.14 to sampleID\n",
      "dtypes: float64(57059), int64(1)\n",
      "memory usage: 4.5+ GB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd0a1a0",
   "metadata": {},
   "source": [
    "## Now we will only keep the genes that are protein coding. Using the ensemble IDs from HGNC we were able to retrieve a dataset that contains only protein coding genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c017ce2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_cd = pd.read_csv('../Data/protein_coding_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "050c33c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl ID(supplied by Ensembl)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000175899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000166535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000184389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Ensembl ID(supplied by Ensembl)\n",
       "0                 ENSG00000121410\n",
       "1                 ENSG00000148584\n",
       "2                 ENSG00000175899\n",
       "3                 ENSG00000166535\n",
       "4                 ENSG00000184389"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_cd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "78932fd7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 19148 entries, 0 to 19147\n",
      "Data columns (total 1 columns):\n",
      " #   Column                           Non-Null Count  Dtype \n",
      "---  ------                           --------------  ----- \n",
      " 0   Ensembl ID(supplied by Ensembl)  19148 non-null  object\n",
      "dtypes: object(1)\n",
      "memory usage: 149.7+ KB\n"
     ]
    }
   ],
   "source": [
    "protein_cd.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b3f5f9cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_to_keep = list(protein_cd['Ensembl ID(supplied by Ensembl)']) #make a list containing protein coding genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9ceb5ddd",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-38-0237e72ac8e7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mall_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "all_cols = list(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "c42361fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSG00000000003'"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_cols[0].split('.')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "51e55dac",
   "metadata": {},
   "outputs": [],
   "source": [
    "our_genes = []\n",
    "for i in range(len(all_cols)):\n",
    "    if all_cols[i].split('.')[0] in genes_to_keep:\n",
    "        our_genes.append(all_cols[i])\n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "a3bfb10b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19037"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(our_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "28ef0f2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_final = df[our_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "0ba5375c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 10530 entries, TCGA-AD-5900-01 to TCGA-FU-A3HZ-01\n",
      "Columns: 19037 entries, ENSG00000000003.14 to ENSG00000282815.1\n",
      "dtypes: float64(19037)\n",
      "memory usage: 1.5+ GB\n"
     ]
    }
   ],
   "source": [
    "df_final.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "abf81fe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_final.to_csv('../Data/data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95bcf6f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## We will now repeat the above procedure for the GTEX samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5f474180",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function for use if we decide to get all of the data\n",
    "num = list(range(0,61,1))\n",
    "\n",
    "names_gtex = ['df'+str(num[x]) for x in range(len(num))]\n",
    "\n",
    "\n",
    "for i in range(len(names_gtex)):\n",
    "    names_gtex[i] = pos(gtex, chunks[i])\n",
    "    names_gtex[i] = names_gtex[i].loc[:, (names_gtex[i]!= 0).any(axis=0)]\n",
    "    names_gtex[i].drop(names_gtex[i].std()[names_gtex[i].std() < 0.5].index.values, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ab2e1802",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes_gtex = []\n",
    "for i in range(len(names_gtex)):\n",
    "    dataframes_gtex.append(names_gtex[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "abd632a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gtex = pd.concat([df.set_index(df.index) for df in dataframes_gtex], ignore_index=False, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "406ee8cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <th>ENSG00000000938.12</th>\n",
       "      <th>ENSG00000000971.15</th>\n",
       "      <th>ENSG00000001036.13</th>\n",
       "      <th>ENSG00000001084.10</th>\n",
       "      <th>ENSG00000001167.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000282785.1</th>\n",
       "      <th>ENSG00000282787.1</th>\n",
       "      <th>ENSG00000282793.1</th>\n",
       "      <th>ENSG00000282795.1</th>\n",
       "      <th>ENSG00000282798.1</th>\n",
       "      <th>ENSG00000282804.1</th>\n",
       "      <th>ENSG00000282807.1</th>\n",
       "      <th>ENSG00000282815.1</th>\n",
       "      <th>ENSG00000282816.1</th>\n",
       "      <th>sampleID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-UTHO-1226-SM-3GAEE</th>\n",
       "      <td>9.124</td>\n",
       "      <td>0.000</td>\n",
       "      <td>9.375</td>\n",
       "      <td>8.514</td>\n",
       "      <td>7.637</td>\n",
       "      <td>10.110</td>\n",
       "      <td>15.720</td>\n",
       "      <td>10.170</td>\n",
       "      <td>8.951</td>\n",
       "      <td>10.190</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.265</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-146FH-1726-SM-5QGQ2</th>\n",
       "      <td>11.360</td>\n",
       "      <td>0.000</td>\n",
       "      <td>9.341</td>\n",
       "      <td>9.245</td>\n",
       "      <td>7.970</td>\n",
       "      <td>7.402</td>\n",
       "      <td>10.620</td>\n",
       "      <td>8.902</td>\n",
       "      <td>10.260</td>\n",
       "      <td>9.446</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.704</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-QDT8-0126-SM-48TZ1</th>\n",
       "      <td>8.926</td>\n",
       "      <td>6.679</td>\n",
       "      <td>9.009</td>\n",
       "      <td>9.016</td>\n",
       "      <td>7.294</td>\n",
       "      <td>9.091</td>\n",
       "      <td>12.820</td>\n",
       "      <td>9.356</td>\n",
       "      <td>9.262</td>\n",
       "      <td>9.365</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.122</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.573</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.122</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-QCQG-1326-SM-48U24</th>\n",
       "      <td>9.862</td>\n",
       "      <td>1.436</td>\n",
       "      <td>9.138</td>\n",
       "      <td>8.547</td>\n",
       "      <td>7.232</td>\n",
       "      <td>8.685</td>\n",
       "      <td>7.618</td>\n",
       "      <td>10.200</td>\n",
       "      <td>9.925</td>\n",
       "      <td>9.481</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.491</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.694</td>\n",
       "      <td>0.8901</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-WZTO-2926-SM-3NM9I</th>\n",
       "      <td>8.429</td>\n",
       "      <td>2.591</td>\n",
       "      <td>9.153</td>\n",
       "      <td>8.497</td>\n",
       "      <td>6.185</td>\n",
       "      <td>7.117</td>\n",
       "      <td>9.470</td>\n",
       "      <td>8.746</td>\n",
       "      <td>9.765</td>\n",
       "      <td>9.334</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.799</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10534</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 55539 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ENSG00000000003.14  ENSG00000000005.5  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                9.124              0.000   \n",
       "GTEX-146FH-1726-SM-5QGQ2              11.360              0.000   \n",
       "GTEX-QDT8-0126-SM-48TZ1                8.926              6.679   \n",
       "GTEX-QCQG-1326-SM-48U24                9.862              1.436   \n",
       "GTEX-WZTO-2926-SM-3NM9I                8.429              2.591   \n",
       "\n",
       "                          ENSG00000000419.12  ENSG00000000457.13  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                9.375               8.514   \n",
       "GTEX-146FH-1726-SM-5QGQ2               9.341               9.245   \n",
       "GTEX-QDT8-0126-SM-48TZ1                9.009               9.016   \n",
       "GTEX-QCQG-1326-SM-48U24                9.138               8.547   \n",
       "GTEX-WZTO-2926-SM-3NM9I                9.153               8.497   \n",
       "\n",
       "                          ENSG00000000460.16  ENSG00000000938.12  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                7.637              10.110   \n",
       "GTEX-146FH-1726-SM-5QGQ2               7.970               7.402   \n",
       "GTEX-QDT8-0126-SM-48TZ1                7.294               9.091   \n",
       "GTEX-QCQG-1326-SM-48U24                7.232               8.685   \n",
       "GTEX-WZTO-2926-SM-3NM9I                6.185               7.117   \n",
       "\n",
       "                          ENSG00000000971.15  ENSG00000001036.13  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE               15.720              10.170   \n",
       "GTEX-146FH-1726-SM-5QGQ2              10.620               8.902   \n",
       "GTEX-QDT8-0126-SM-48TZ1               12.820               9.356   \n",
       "GTEX-QCQG-1326-SM-48U24                7.618              10.200   \n",
       "GTEX-WZTO-2926-SM-3NM9I                9.470               8.746   \n",
       "\n",
       "                          ENSG00000001084.10  ENSG00000001167.14  ...  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                8.951              10.190  ...   \n",
       "GTEX-146FH-1726-SM-5QGQ2              10.260               9.446  ...   \n",
       "GTEX-QDT8-0126-SM-48TZ1                9.262               9.365  ...   \n",
       "GTEX-QCQG-1326-SM-48U24                9.925               9.481  ...   \n",
       "GTEX-WZTO-2926-SM-3NM9I                9.765               9.334  ...   \n",
       "\n",
       "                          ENSG00000282785.1  ENSG00000282787.1  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                 0.0              0.000   \n",
       "GTEX-146FH-1726-SM-5QGQ2                0.0              0.000   \n",
       "GTEX-QDT8-0126-SM-48TZ1                 0.0              2.122   \n",
       "GTEX-QCQG-1326-SM-48U24                 0.0              3.491   \n",
       "GTEX-WZTO-2926-SM-3NM9I                 0.0              0.000   \n",
       "\n",
       "                          ENSG00000282793.1  ENSG00000282795.1  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                 0.0                0.0   \n",
       "GTEX-146FH-1726-SM-5QGQ2                0.0                0.0   \n",
       "GTEX-QDT8-0126-SM-48TZ1                 0.0                0.0   \n",
       "GTEX-QCQG-1326-SM-48U24                 0.0                0.0   \n",
       "GTEX-WZTO-2926-SM-3NM9I                 0.0                0.0   \n",
       "\n",
       "                          ENSG00000282798.1  ENSG00000282804.1  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE               3.265             0.0000   \n",
       "GTEX-146FH-1726-SM-5QGQ2              4.704             0.0000   \n",
       "GTEX-QDT8-0126-SM-48TZ1               3.573             0.0000   \n",
       "GTEX-QCQG-1326-SM-48U24               3.694             0.8901   \n",
       "GTEX-WZTO-2926-SM-3NM9I               3.799             0.0000   \n",
       "\n",
       "                          ENSG00000282807.1  ENSG00000282815.1  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                 0.0              0.000   \n",
       "GTEX-146FH-1726-SM-5QGQ2                0.0              0.000   \n",
       "GTEX-QDT8-0126-SM-48TZ1                 0.0              2.122   \n",
       "GTEX-QCQG-1326-SM-48U24                 0.0              0.000   \n",
       "GTEX-WZTO-2926-SM-3NM9I                 0.0              1.284   \n",
       "\n",
       "                          ENSG00000282816.1  sampleID  \n",
       "GTEX-UTHO-1226-SM-3GAEE                 0.0     10530  \n",
       "GTEX-146FH-1726-SM-5QGQ2                0.0     10531  \n",
       "GTEX-QDT8-0126-SM-48TZ1                 0.0     10532  \n",
       "GTEX-QCQG-1326-SM-48U24                 0.0     10533  \n",
       "GTEX-WZTO-2926-SM-3NM9I                 0.0     10534  \n",
       "\n",
       "[5 rows x 55539 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtex.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5f777f75",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_cols_gtex = list(df_gtex.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "08d383e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "our_genes_gtex = []\n",
    "for i in range(len(all_cols_gtex)):\n",
    "    if all_cols_gtex[i].split('.')[0] in genes_to_keep:\n",
    "        our_genes_gtex.append(all_cols_gtex[i])\n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b7db1996",
   "metadata": {},
   "outputs": [],
   "source": [
    "#again keep only the protein coding genes\n",
    "\n",
    "df_gtex = df_gtex[our_genes_gtex]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "64c6a2e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <th>ENSG00000000938.12</th>\n",
       "      <th>ENSG00000000971.15</th>\n",
       "      <th>ENSG00000001036.13</th>\n",
       "      <th>ENSG00000001084.10</th>\n",
       "      <th>ENSG00000001167.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000280314.1</th>\n",
       "      <th>ENSG00000280670.2</th>\n",
       "      <th>ENSG00000280789.1</th>\n",
       "      <th>ENSG00000280969.1</th>\n",
       "      <th>ENSG00000281106.2</th>\n",
       "      <th>ENSG00000281991.1</th>\n",
       "      <th>ENSG00000282419.1</th>\n",
       "      <th>ENSG00000282608.1</th>\n",
       "      <th>ENSG00000282757.1</th>\n",
       "      <th>ENSG00000282815.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-UTHO-1226-SM-3GAEE</th>\n",
       "      <td>9.124</td>\n",
       "      <td>0.000</td>\n",
       "      <td>9.375</td>\n",
       "      <td>8.514</td>\n",
       "      <td>7.637</td>\n",
       "      <td>10.110</td>\n",
       "      <td>15.720</td>\n",
       "      <td>10.170</td>\n",
       "      <td>8.951</td>\n",
       "      <td>10.190</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.260</td>\n",
       "      <td>9.723</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>6.241</td>\n",
       "      <td>4.746</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.460</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-146FH-1726-SM-5QGQ2</th>\n",
       "      <td>11.360</td>\n",
       "      <td>0.000</td>\n",
       "      <td>9.341</td>\n",
       "      <td>9.245</td>\n",
       "      <td>7.970</td>\n",
       "      <td>7.402</td>\n",
       "      <td>10.620</td>\n",
       "      <td>8.902</td>\n",
       "      <td>10.260</td>\n",
       "      <td>9.446</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.454</td>\n",
       "      <td>9.839</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2.861</td>\n",
       "      <td>8.651</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.521</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-QDT8-0126-SM-48TZ1</th>\n",
       "      <td>8.926</td>\n",
       "      <td>6.679</td>\n",
       "      <td>9.009</td>\n",
       "      <td>9.016</td>\n",
       "      <td>7.294</td>\n",
       "      <td>9.091</td>\n",
       "      <td>12.820</td>\n",
       "      <td>9.356</td>\n",
       "      <td>9.262</td>\n",
       "      <td>9.365</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.563</td>\n",
       "      <td>9.641</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2.780</td>\n",
       "      <td>8.876</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.825</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-QCQG-1326-SM-48U24</th>\n",
       "      <td>9.862</td>\n",
       "      <td>1.436</td>\n",
       "      <td>9.138</td>\n",
       "      <td>8.547</td>\n",
       "      <td>7.232</td>\n",
       "      <td>8.685</td>\n",
       "      <td>7.618</td>\n",
       "      <td>10.200</td>\n",
       "      <td>9.925</td>\n",
       "      <td>9.481</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.507</td>\n",
       "      <td>10.710</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2.614</td>\n",
       "      <td>7.030</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-WZTO-2926-SM-3NM9I</th>\n",
       "      <td>8.429</td>\n",
       "      <td>2.591</td>\n",
       "      <td>9.153</td>\n",
       "      <td>8.497</td>\n",
       "      <td>6.185</td>\n",
       "      <td>7.117</td>\n",
       "      <td>9.470</td>\n",
       "      <td>8.746</td>\n",
       "      <td>9.765</td>\n",
       "      <td>9.334</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.627</td>\n",
       "      <td>10.450</td>\n",
       "      <td>0.7804</td>\n",
       "      <td>5.850</td>\n",
       "      <td>6.773</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.961</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.284</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19036 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ENSG00000000003.14  ENSG00000000005.5  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                9.124              0.000   \n",
       "GTEX-146FH-1726-SM-5QGQ2              11.360              0.000   \n",
       "GTEX-QDT8-0126-SM-48TZ1                8.926              6.679   \n",
       "GTEX-QCQG-1326-SM-48U24                9.862              1.436   \n",
       "GTEX-WZTO-2926-SM-3NM9I                8.429              2.591   \n",
       "\n",
       "                          ENSG00000000419.12  ENSG00000000457.13  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                9.375               8.514   \n",
       "GTEX-146FH-1726-SM-5QGQ2               9.341               9.245   \n",
       "GTEX-QDT8-0126-SM-48TZ1                9.009               9.016   \n",
       "GTEX-QCQG-1326-SM-48U24                9.138               8.547   \n",
       "GTEX-WZTO-2926-SM-3NM9I                9.153               8.497   \n",
       "\n",
       "                          ENSG00000000460.16  ENSG00000000938.12  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                7.637              10.110   \n",
       "GTEX-146FH-1726-SM-5QGQ2               7.970               7.402   \n",
       "GTEX-QDT8-0126-SM-48TZ1                7.294               9.091   \n",
       "GTEX-QCQG-1326-SM-48U24                7.232               8.685   \n",
       "GTEX-WZTO-2926-SM-3NM9I                6.185               7.117   \n",
       "\n",
       "                          ENSG00000000971.15  ENSG00000001036.13  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE               15.720              10.170   \n",
       "GTEX-146FH-1726-SM-5QGQ2              10.620               8.902   \n",
       "GTEX-QDT8-0126-SM-48TZ1               12.820               9.356   \n",
       "GTEX-QCQG-1326-SM-48U24                7.618              10.200   \n",
       "GTEX-WZTO-2926-SM-3NM9I                9.470               8.746   \n",
       "\n",
       "                          ENSG00000001084.10  ENSG00000001167.14  ...  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                8.951              10.190  ...   \n",
       "GTEX-146FH-1726-SM-5QGQ2              10.260               9.446  ...   \n",
       "GTEX-QDT8-0126-SM-48TZ1                9.262               9.365  ...   \n",
       "GTEX-QCQG-1326-SM-48U24                9.925               9.481  ...   \n",
       "GTEX-WZTO-2926-SM-3NM9I                9.765               9.334  ...   \n",
       "\n",
       "                          ENSG00000280314.1  ENSG00000280670.2  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                 0.0              5.260   \n",
       "GTEX-146FH-1726-SM-5QGQ2                0.0              6.454   \n",
       "GTEX-QDT8-0126-SM-48TZ1                 0.0              4.563   \n",
       "GTEX-QCQG-1326-SM-48U24                 0.0              7.507   \n",
       "GTEX-WZTO-2926-SM-3NM9I                 0.0              7.627   \n",
       "\n",
       "                          ENSG00000280789.1  ENSG00000280969.1  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE               9.723             0.0000   \n",
       "GTEX-146FH-1726-SM-5QGQ2              9.839             0.0000   \n",
       "GTEX-QDT8-0126-SM-48TZ1               9.641             0.0000   \n",
       "GTEX-QCQG-1326-SM-48U24              10.710             0.0000   \n",
       "GTEX-WZTO-2926-SM-3NM9I              10.450             0.7804   \n",
       "\n",
       "                          ENSG00000281106.2  ENSG00000281991.1  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE               6.241              4.746   \n",
       "GTEX-146FH-1726-SM-5QGQ2              2.861              8.651   \n",
       "GTEX-QDT8-0126-SM-48TZ1               2.780              8.876   \n",
       "GTEX-QCQG-1326-SM-48U24               2.614              7.030   \n",
       "GTEX-WZTO-2926-SM-3NM9I               5.850              6.773   \n",
       "\n",
       "                          ENSG00000282419.1  ENSG00000282608.1  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE                 0.0              7.460   \n",
       "GTEX-146FH-1726-SM-5QGQ2                0.0              5.521   \n",
       "GTEX-QDT8-0126-SM-48TZ1                 0.0              6.825   \n",
       "GTEX-QCQG-1326-SM-48U24                 0.0              5.300   \n",
       "GTEX-WZTO-2926-SM-3NM9I                 0.0              4.961   \n",
       "\n",
       "                          ENSG00000282757.1  ENSG00000282815.1  \n",
       "GTEX-UTHO-1226-SM-3GAEE                 0.0              0.000  \n",
       "GTEX-146FH-1726-SM-5QGQ2                0.0              0.000  \n",
       "GTEX-QDT8-0126-SM-48TZ1                 0.0              2.122  \n",
       "GTEX-QCQG-1326-SM-48U24                 0.0              0.000  \n",
       "GTEX-WZTO-2926-SM-3NM9I                 0.0              1.284  \n",
       "\n",
       "[5 rows x 19036 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtex.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b7b343e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 7775 entries, GTEX-UTHO-1226-SM-3GAEE to GTEX-12BJ1-0426-SM-5FQSO\n",
      "Columns: 19036 entries, ENSG00000000003.14 to ENSG00000282815.1\n",
      "dtypes: float64(19036)\n",
      "memory usage: 1.1+ GB\n"
     ]
    }
   ],
   "source": [
    "df_gtex.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "f05d5ac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gtex.to_csv('../Data/df_gtex.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79e25726",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
